File: src/main/java/komu/blunt/eval/Evaluator.java
package komu.blunt.eval;

import com.google.common.io.Resources;
import komu.blunt.analyzer.Analyzer;
import komu.blunt.analyzer.StaticEnvironment;
import komu.blunt.analyzer.VariableReference;
import komu.blunt.asm.*;
import komu.blunt.ast.*;
import komu.blunt.core.CoreDefineExpression;
import komu.blunt.core.CoreExpression;
import komu.blunt.parser.Parser;
import komu.blunt.stdlib.BasicFunctions;
import komu.blunt.types.*;
import komu.blunt.types.checker.TypeChecker;

import java.io.IOException;
import java.net.URL;
import java.nio.charset.Charset;

import static komu.blunt.eval.ConstructorArgumentCollector.createConstructor;
import static komu.blunt.types.Predicate.isIn;

public final class Evaluator {

    private final RootBindings rootBindings = new RootBindings();
    private final Instructions instructions = new Instructions();
    private final ClassEnv classEnv = new ClassEnv();
    private long steps = 0;

    public Evaluator() {
        new NativeFunctionRegisterer(rootBindings).register(BasicFunctions.class);

        for (ConstructorDefinition constructor : rootBindings.dataTypes.getDeclaredConstructors())
            createConstructorFunction(constructor);
    }

    public CoreExpression analyze(ASTExpression expr) {
        typeCheck(expr);
        CoreExpression exp = toCore(expr, rootBindings.staticEnvironment.extend());
        return exp;
    }
    
    public void load(String source) {
        Parser parser = new Parser(source);

        MyDefinitionVisitor visitor = new MyDefinitionVisitor();
        for (ASTDefinition define : parser.parseDefinitions())
            define.accept(visitor, null);
    }

    private final class MyDefinitionVisitor implements ASTDefinitionVisitor<Void,Void> {
        @Override
        public Void visit(ASTValueDefinition definition, Void ctx) {
            Scheme type = TypeChecker.typeCheck(definition, classEnv, rootBindings.dataTypes, rootBindings.createAssumptions());
            
            rootBindings.defineVariableType(definition.name, type);
            VariableReference var = rootBindings.staticEnvironment.define(definition.name);

            CoreExpression exp = Analyzer.analyze(definition.value, rootBindings.dataTypes, rootBindings.staticEnvironment);

            run(new CoreDefineExpression(exp, var), rootBindings.runtimeEnvironment);
            return null;
        }

        @Override
        public Void visit(ASTDataDefinition definition, Void ctx) {
            register(definition);

            return null;
        }
    }

    private void register(ASTDataDefinition definition) {
        rootBindings.dataTypes.register(definition);

        for (ConstructorDefinition ctor : definition.constructors)
            createConstructorFunction(ctor);
        
        for (String className : definition.derivedClasses) {
            classEnv.addInstance(isIn(className, definition.type));
        }
    }

    private void createConstructorFunction(ConstructorDefinition ctor) {
        if (ctor.arity != 0)
            rootBindings.bind(ctor.name, ctor.scheme, createConstructor(ctor));
    }

    private Object run(CoreExpression expression, Environment env) {
        int pos = instructions.pos();

        instructions.append(expression.simplify().assemble(new Assembler(), Register.VAL, Linkage.NEXT));

        VM vm = new VM(instructions, env, rootBindings.runtimeEnvironment);
        vm.set(Register.PC, pos);
        Object result = vm.run();
        steps += vm.steps;
        return result;
    }

    public Object evaluate(ASTExpression exp) {
        return evaluateWithType(exp).result;
    }

    public ResultWithType evaluateWithType(ASTExpression exp) {
        StaticEnvironment env = rootBindings.staticEnvironment.extend();
        Qualified<Type> type = typeCheck(exp);
        CoreExpression expression = toCore(exp, env);

        Object result = run(expression, rootBindings.runtimeEnvironment.extend(env.size()));

        return new ResultWithType(result, type);
    }

    private Qualified<Type> typeCheck(ASTExpression exp) {
        return TypeChecker.typeCheck(exp, classEnv, rootBindings.dataTypes, rootBindings.createAssumptions());
    }

    private CoreExpression toCore(ASTExpression exp, StaticEnvironment env) {
        return Analyzer.analyze(exp, rootBindings.dataTypes, env);
    }

    public long getSteps() {
        return steps;
    }

    public void dump() {
        instructions.dump();
    }

    public void loadResource(String path) throws IOException {
        load(readResource(path));
    }

    private String readResource(String path) throws IOException {
        URL resource = getClass().getClassLoader().getResource(path);
        return Resources.toString(resource, Charset.forName("UTF-8"));
    }
}

File: src/main/kotlin/komu/blunt/eval/Evaluator.kt
package komu.blunt.eval

import komu.blunt.analyzer.StaticEnvironment
import komu.blunt.analyzer.analyze
import komu.blunt.asm.Assembler
import komu.blunt.asm.Instructions
import komu.blunt.asm.Register
import komu.blunt.asm.VM
import komu.blunt.ast.ASTDefinition
import komu.blunt.ast.ASTExpression
import komu.blunt.core.CoreDefineExpression
import komu.blunt.core.CoreExpression
import komu.blunt.parser.Parser
import komu.blunt.stdlib.BasicFunctions
import komu.blunt.types.*
import komu.blunt.types.checker.TypeChecker
import komu.blunt.utils.readResourceAsString

class Evaluator() {

    private val rootBindings = RootBindings()
    private val instructions = Instructions()
    private val classEnv = ClassEnv()
    var steps = 0.toLong();

    init {
        BasicFunctions.register(rootBindings)

        for (constructor in rootBindings.dataTypes.declaredConstructors)
            createConstructorFunction(constructor)
    }

    fun analyze(expr: ASTExpression): CoreExpression {
        typeCheck(expr)
        return toCore(expr, rootBindings.staticEnvironment.extend())
    }

    fun load(source: String) {
        val parser = Parser(source)

        for (define in parser.parseDefinitions())
            when (define) {
                is ASTDefinition.Value -> processDefinition(define)
                is ASTDefinition.Data  -> register(define)
            }
    }

    private fun processDefinition(definition: ASTDefinition.Value) {
        try {
            val typ = TypeChecker.typeCheck(definition, classEnv, rootBindings.dataTypes, rootBindings.createAssumptions())

            rootBindings.defineVariableType(definition.name, typ)
            val v = rootBindings.staticEnvironment.define(definition.name)

            val exp = analyze(definition.value, rootBindings.dataTypes, rootBindings.staticEnvironment)

            run(CoreDefineExpression(v, exp), rootBindings.runtimeEnvironment)

        } catch (e: TypeCheckException) {
            println("failure in definition: $definition")
            throw e
        }
    }

    private fun register(definition: ASTDefinition.Data) {
        rootBindings.dataTypes.register(definition)

        for (ctor in definition.constructors)
            createConstructorFunction(ctor)

        for (className in definition.derivedClasses)
            classEnv.addInstance(isIn(className, definition.type))
    }

    private fun createConstructorFunction(ctor: ConstructorDefinition) {
        if (ctor.arity != 0)
            rootBindings.bind(ctor.name, ctor.scheme, createConstructor(ctor));
    }

    private fun run(expression: CoreExpression, env: Environment): Any? {
        val pos = instructions.count

        instructions += expression.simplify().assemble(Assembler(), Register.VAL)

        val vm = VM(instructions, env, rootBindings.runtimeEnvironment)
        vm.pc = pos
        val result = vm.run()
        steps += vm.steps
        return result
    }

    fun evaluate(exp: ASTExpression): Any? =
        evaluateWithType(exp).first

    fun evaluateWithType(exp: ASTExpression): Pair<Any?, Qualified<Type>> {
        val env = rootBindings.staticEnvironment.extend()
        val typ = typeCheck(exp)
        val expression = toCore(exp, env);

        val result = run(expression, rootBindings.runtimeEnvironment.extend(env.size))

        return Pair(result, typ)
    }

    private fun typeCheck(exp: ASTExpression) =
        TypeChecker.typeCheck(exp, classEnv, rootBindings.dataTypes, rootBindings.createAssumptions())

    private fun toCore(exp: ASTExpression, env: StaticEnvironment) =
        analyze(exp, rootBindings.dataTypes, env)

    fun dump() {
        instructions.dump()
    }

    fun loadResource(path: String) {
        load(readResource(path))
    }

    private fun readResource(path: String): String =
        javaClass.getClassLoader()!!.readResourceAsString(path) ?:
            throw Exception("could not find resource: $path")
}
